### ISSUE TITLE
Getting 405 Error when running the notebook Usage_example_FDP.ipynb

### ISSUE THREAD

**Issue by priya-gitTest** [Health-RI/SeMPyRO#9]
![image](https://github.com/Health-RI/SeMPyRO/assets/21082240/1607df50-2945-4bab-b01f-c818b8c7d592)
![image](https://github.com/Health-RI/SeMPyRO/assets/21082240/cd07232b-ce83-4093-aa28-53e5de0db043)
--
400 Client Error:  for url: http://localhost//catalog
@prefix rsx: <http://rdf4j.org/shacl-extensions#> .
@prefix rdf4j: <http://rdf4j.org/schema/rdf4j#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix dcterms: <http://purl.org/dc/terms/> .
@prefix dcat: <http://www.w3.org/ns/dcat#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix ldp: <http://www.w3.org/ns/ldp#> .

[] a sh:ValidationReport;
  sh:conforms false;
  rdf4j:truncated false;
  sh:result [ a sh:ValidationResult;
      sh:focusNode <http://localhost:8080/catalog/bb046e9f-b09a-4ed5-b7be-d6130fcd87b4>;
      rsx:shapesGraph rdf4j:SHACLShapeGraph;
      sh:value <https://harrypotter.fandom.com/wiki/Hogwarts_School_of_Witchcraft_and_Wizardry>;
      sh:resultPath dcterms:publisher;
      sh:sourceConstraintComponent sh:NodeConstraintComponent;
      sh:resultSeverity sh:Violation;
      sh:sourceShape [ a sh:PropertyShape;
          sh:path dcterms:publisher;
          sh:node <http://fairdatapoint.org/AgentShape>
        ]
    ], [ a sh:ValidationResult;
      sh:focusNode <http://localhost:8080/catalog/bb046e9f-b09a-4ed5-b7be-d6130fcd87b4>;
      rsx:shapesGraph rdf4j:SHACLShapeGraph;
      sh:resultPath dcterms:hasVersion;
      sh:sourceConstraintComponent sh:MinCountConstraintComponent;
      sh:resultSeverity sh:Violation;
      sh:sourceShape [ a sh:PropertyShape;
          sh:path dcterms:hasVersion;
          sh:minCount 1
        ]
    ] .

<http://fairdatapoint.org/AgentShape> a sh:NodeShape;
  sh:property [ a sh:PropertyShape;
      sh:path foaf:name;
      sh:minCount 1;
      sh:maxCount 1;
      sh:nodeKind sh:Literal
    ] .

**Comment by a-nayden** [comment]
Hi @priya-gitTest ,
Looks like base FDP link provided at login missing port after localhost. Could you double-check please the link you are providing at login? Is it just `localhost` or `localhost:8080`?

**Comment by Markus92** [comment]
400 client error could also be caused by trailing slash in FDP uri, as I see http://localhost//catalog instead of http://localhost/catalog

**Comment by priya-gitTest** [comment]
The error is still the same without the double//. I think, I might know the issue but will let you know once resolved . The problem is that the base_url is assume to be at port 8080 and seems mines in at port 80. Need to figure out, where to fix it but seems hardcoded in your code at least somewhere.

**Comment by priya-gitTest** [comment]
still not resolved :( 
![image](https://github.com/Health-RI/SeMPyRO/assets/21082240/76da8933-e024-4bbe-8377-2553748c24c0)

**Comment by priya-gitTest** [comment]
Any Updates for this error ? I tried today again using the latest code and your Test URL [https://fdp-test.healthdata.nl/] for user : [nikola.tesla@example.com](mailto:nikola.tesla%40example.com) but no luck. 

![image](https://github.com/user-attachments/assets/95e2fba8-7aee-4ea3-9fc4-12208ef7f5f5)

**Comment by Markus92** [comment]
There's a trailing / in your FDP url. Make sure base_uri is without a trailing slash.

**Comment by priya-gitTest** [comment]
> There's a trailing / in your FDP url. Make sure base_uri is without a trailing slash.

yes, tried again now with the right URL. Still the same results :( 

![image](https://github.com/user-attachments/assets/e9e6bd93-08f2-4bb1-9635-706886163e4f)

**Comment by Markus92** [comment]
@priya-gitTest I went through the example and updated it, cam you check https://github.com/Health-RI/SeMPyRO/pull/26 ?

You'll have to use your own local install as it is made for the default shapes. But the default docker compose file from the documentation should work fine (it's what I tested against).

**Comment by priya-gitTest** [comment]
> @priya-gitTest I went through the example and updated it, can you check #26 ?
> 
> You'll have to use your own local install as it is made for the default shapes. But the [default docker compose file](https://fairdatapoint.readthedocs.io/en/latest/deployment/local-deployment.html) from the documentation should work (it's what I tested against).

I will check and let you know. I made this code work for the step of adding the Catalog and that goes fine. Will share the code later. Was trying to fix the DataSet as I am not stuck there with SHACL Errors : 

E.g. 
from rdflib import DCTERMS, RDF,DCAT,XSD
from urllib.parse import quote
from datetime import datetime

def create_agent(graph : Graph, uri :str, name :str, identifier:str) ->URIRef:
    agent = URIRef(uri)
    graph.add((agent, RDF.type, FOAF.Agent))
    graph.add((agent, FOAF.name, Literal(name)))
    graph.add((agent, DCTERMS.identifier, Literal(identifier)))
    
    return agent

def create_dataset_graph(record:Dict)->Graph:
    g = Graph()

    dataset_uri =  URIRef(f"http://example.com/dataset_{record['id'][0]}")

    #Create and add creator
    creator = create_agent(graph,URIRef(record["author_id"]),record["author_name"],record["author_id"])

    #Create and add publisher
    publisher = create_agent(graph,URIRef(record["publisher_id"]),record["publisher_name"],record["publisher_id"] )

    issued_date = datetime.strptime(record["issued"],'%d-%m-%Y').isoformat()
    g.add((dataset_uri,RDF.type,DCAT.Dataset))
    g.add((dataset_uri,DCTERMS.title,Literal(record["name"])))
    g.add((dataset_uri,DCTERMS.description, Literal(record["description"])))
    g.add((dataset_uri,DCTERMS.identifier,Literal(record["id"][0])))
    g.add((dataset_uri,DCTERMS.isPartOf,URIRef(catalog_fdp_id)))
    g.add((dataset_uri,DCTERMS.issued,Literal(issued_date,datatype=XSD.dateTime)))
    g.add((dataset_uri,DCTERMS.creator,creator))
    g.add((dataset_uri,DCTERMS.publisher,publisher))

    for keyword in record["keywords"]:
        g.add((dataset_uri,DCAT.keyword,Literal(keyword.strip())))

    for theme in record["theme"]:
        g.add((dataset_uri,DCAT.theme,URIRef(quote(theme.strip(),safe=":/"))))

    return g 


datasets = df.to_dict('records')
for record in datasets:
    #print(record)
    #Create a graph to define publisher as a RDF resource
    dataset_graph = create_dataset_graph(record)
    print(dataset_graph.serialize())#format='turtle'))
    try:
        #fdp_client = FDPClient(base_url=fdp_base, username=username, password=password, verify_ssl=True)
        data_fdp_id = fdp_client.create_and_publish(resource_type="dataset", metadata=dataset_graph)
        print(data_fdp_id)
    except FDPClientException as e:
        print(f"An error occured : {str(e)}")

**Comment by Markus92** [comment]
Ah got ya. Btw might not be the best idea to test against our semi-private test instance, we play around with it a lot.

**Comment by priya-gitTest** [comment]
> Ah got ya. Btw might not be the best idea to test against our semi-private test instance, we play around with it a lot.

Ok ok. I hope its just some tweaking for this code, to insert the dataset as In my local instance I could do that using my own code. But the issue is the SHACL validation and and somehow need to play around it anyways to understand it :) .

**Comment by priya-gitTest** [comment]
@Markus92 It  works now: check for this record : https://fdp-test.healthdata.nl/catalog/81cbb7da-eb9d-4a00-914e-3aea5aaccf06

You can delete other records if possible. I will clean the code and share with you soon !!!
